core/src/main/scala/org/apache/spark/deploy/SparkHadoopUtil.scala

```diff
@@ -353,6 +354,25 @@ class SparkHadoopUtil extends Logging {
     }
     buffer.toString
   }
+
+  /**
+   * Run some code as the real logged in user (which may differ from the current user, for
+   * example, when using proxying).
+   */
+  private[spark] def doAsRealUser[T](fn: => T): T = {
+    val currentUser = UserGroupInformation.getCurrentUser()
```

Comment:
Hmmm... I'm not so sure this will work in all cases. Can you test this with both `spark.sql.hive.metastore.jars` and `spark.sql.hive.metastore.version` set?

The problem is that this class is loaded by Spark's main class loader, while `HiveClientImpl` comes from a different class loader. So `UserGroupInformation` might be a different class in certain cases. It's the same reasoning why `HiveClientImpl` class does its own `loginUserFromKeytab` around L110.